README.md
# puppet

#### Table of Contents

1. [Overview](#overview)
2. [Module Description - What the module does and why it is useful](#module-description)
3. [Setup - The basics of getting started with puppet](#setup)
    * [What puppet affects](#what-puppet-affects)
    * [Beginning with puppet](#beginning-with-puppet)
4. [Usage - Configuration options and additional functionality](#usage)
5. [Reference - An under-the-hood peek at what the module is doing and how](#reference)
6. [Development - Guide for contributing to the module](#development)

## Overview

puppet agent management

## Module Description

If applicable, this section should have a brief description of the technology
the module integrates with and what that integration enables. This section
should answer the questions: "What does this module *do*?" and "Why would I use
it?"

If your module has a range of functionality (installation, configuration,
management, etc.) this is the time to mention it.

## Setup

### What puppet affects

* A list of files, packages, services, or operations that the module will alter,
  impact, or execute on the system it's installed on.
* This is a great place to stick any warnings.
* Can be in list or paragraph form.

### Beginning with puppet

```
class { 'puppet::agent':
  puppetmaster       => 'demoserver',
  puppetmasterport   => '1234',
  puppetenv          => 'tstenv',
}
```

## Usage

### multiple agents

To be able to have multiple agents we can use a cronjobs:

```
puppet::agent::cron { 'puppet.systemadmin.es':
  masterport => '8140',
}
```

## Reference

Here, list the classes, types, providers, facts, etc contained in your module.
This section should include all of the under-the-hood workings of your module so
people know what the module is touching on their system but don't need to mess
with things. (We are working on automating this section!)


## Development

We are pushing to have acceptance testing in place, so any new feature should
have some tests to check both presence and absence of any feature

### TODO

* improve configuration options

### Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
# CHANGELOG

## 0.3.18
* added error message case for "Failed to apply catalog"

## 0.3.17

* added different error codes for each error on puppet_last_run.sh
* added chek to review if puppet agent was manualy disabled on puppet_last_run.sh

## 0.3.16

* added support for **Ubuntu 20.04**

## 0.3.14

* added description for cron-based and agent base agents

## 0.3.13

* Added resource report to snmpd's puppetlr check
* Improved **check_last_puppet_run** performance data report

## 0.3.12

* Improved **check_last_puppet_run**
* Added snmpd compatible last run check

## 0.3.10

* added support for **RHEL 8**

## 0.3.9

* Added ssldir option to **puppet::agent::cron**

## 0.3.8

* bugfix for **Debian 10**/**Rapbian 10**

## 0.3.5

* added support for **Debian 10**

## 0.3.4

* added **puppet::agent::cron**

## 0.3.3

* added **noop** to **puppet::agent**

## 0.3.2

* added runinterval setting for **puppet::agent**

## 0.3.1

* bugfix

## 0.3.0

* environment variable for puppet agent

## 0.2.3

* added support for **SLES11SP4**

## 0.2.2

* added support for **SLES11SP3**

## 0.2.1

* added option to disable management of puppet agent's configuration

## 0.2.0

* updated to **puppet 5**
* Added SLES12SP3 support
* **puppet::client** renamed to **puppet::agent**
* added **puppet::client** compatibility class that translats variables to **puppet::agent**

## 0.1.27

* added Ubuntu 18.04 support

## 0.1.26

* improved nagios check: **check_last_puppet_run** and moved to the files directory

## 0.1.25

* added service_enable varible to **puppet::client**

## 0.1.24

* Ubuntu 16.04 support

## 0.1.23

* bugfix **check_last_puppet_run**

## 0.1.22

* added **check_last_puppet_run**


spec/acceptance/version.rb

_osfamily               = fact('osfamily')
_operatingsystem        = fact('operatingsystem')
_operatingsystemrelease = fact('operatingsystemrelease').to_f

case _osfamily
when 'RedHat'
  $packagename  = ''

when 'Debian'
  $packagename = ''

else
  $packagename = '-_-'

end


spec/acceptance/base_spec.rb
require 'spec_helper_acceptance'
require_relative './version.rb'

describe 'puppet class' do

  context 'basic setup' do
    # Using puppet_apply as a helper
    it 'should work with no errors' do
      pp = <<-EOF

      class { 'puppet::agent':
        puppetmaster     => 'lolmaster',
        puppetmasterport => '1234',
      }

      EOF

      # Run it twice and test for idempotency
      expect(apply_manifest(pp).exit_code).to_not eq(1)
      expect(apply_manifest(pp).exit_code).to eq(0)
    end

    #/etc/puppetlabs/puppet/puppet.conf
    # server = <%= @puppetmaster %>
    # masterport = <%= @puppetmasterport %>
    describe file('/etc/puppetlabs/puppet/puppet.conf') do
      it { should be_file }
      its(:content) { should match 'puppet managed file' }
      its(:content) { should match 'server = lolmaster' }
      its(:content) { should match 'masterport = 1234' }
    end

    it "env production" do
      expect(shell("puppet config print environment | grep production").exit_code).to be_zero
    end

    it 'tstenv should work with no errors' do
      pp = <<-EOF

      class { 'puppet::agent':
        puppetmaster       => 'lolmaster',
        puppetmasterport   => '1234',
        puppetenv          => 'tstenv',
      }

      EOF

      # Run it twice and test for idempotency
      expect(apply_manifest(pp).exit_code).to_not eq(1)
    end

    describe file('/etc/puppetlabs/puppet/puppet.conf') do
      it { should be_file }
      its(:content) { should match 'puppet managed file' }
      its(:content) { should match 'server = lolmaster' }
      its(:content) { should match 'masterport = 1234' }
      its(:content) { should match 'environment=tstenv' }
    end

    it "puppet configured env" do
      expect(shell("puppet config print environment").exit_code).to be_zero
    end

    it "env tst_env" do
      expect(shell("puppet config print environment | grep tstenv").exit_code).to be_zero
    end

  end
end


files/check_last_puppet_run.sh
#!/bin/bash
# puppet managed file

PATH="/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/opt/puppetlabs/bin"

MAXDIFF=7200
WARNDIFF=3600

while getopts 'c:w:' OPT; do
  case $OPT in
    c)  MAXDIFF=$OPTARG;;
    w)  WARNDIFF=$OPTARG;;
  esac
done

shift $(($OPTIND - 1))

PUPPETBIN=$(which puppet 2>/dev/null)

if [ -z "${PUPPETBIN}" ];
then
  echo "CRITICAL - puppet not found"
  exit 2
fi

PUPPET_VER="$(${PUPPETBIN} --version 2>/dev/null)"

if [[ $PUPPET_VER = 3* ]];
then
  LAST_RUN_FILE='/var/lib/puppet/state/last_run_summary.yaml'
  LAST_RUN_REPORT='/var/lib/puppet/state/last_run_report.yaml'
elif [[ $PUPPET_VER = 5* ]];
then
  LAST_RUN_FILE='/opt/puppetlabs/puppet/cache/state/last_run_summary.yaml'
  LAST_RUN_REPORT='/opt/puppetlabs/puppet/cache/state/last_run_report.yaml'
else
  echo "CRITICAL - Unsupported puppet version"
  exit 2
fi

if [ ! -e "${LAST_RUN_FILE}" ];
then
	echo "CRITICAL - ${LAST_RUN_FILE} does not exists"
	exit 2
fi

if [ ! -e "${LAST_RUN_REPORT}" ];
then
	echo "CRITICAL - ${LAST_RUN_REPORT} does not exists"
	exit 2
fi

PERFDATA="$PERFDATA $(grep resources: ${LAST_RUN_FILE} -A7 | grep -v resources: | paste '-sd;' | sed -e 's/: /=/g' -e 's/^[^a-zA-Z]*//' -e 's/[ \t]+/ /g')"

#
# outputs
#
# # grep -i catalo /var/lib/puppet/state/last_run_report.yaml
#       message: "Could not retrieve catalog from remote server: Error 400 on SERVER: Puppet::Parser::AST::Resource failed with error ArgumentError: Could not find declared class sssd::monit at /etc/puppet/manifests/site.pp:919 on node testvm.systemadmin.es"
#       message: "Using cached catalog"
#       message: "Finished catalog run in 7.96 seconds"
#

grep "Using cached catalog" "${LAST_RUN_REPORT}" >/dev/null 2>&1
if [ "$?" -eq 0 ];
then
  echo "CRITICAL - server using cached catalog |${PERFDATA}"
  exit 2
fi

# # grep -i catalo /var/lib/puppet/state/last_run_report.yaml
#       message: "Could not retrieve catalog from remote server: Error 400 on SERVER: Puppet::Parser::AST::Resource failed with error ArgumentError: Invalid resource type pam::securetty at /etc/puppet/manifests/site.pp:356 on node demovm.systemadmin.es"
#       message: "Not using cache on failed catalog"
#       message: "Could not retrieve catalog; skipping run"
#
grep "Could not retrieve catalog from remote server" ${LAST_RUN_REPORT} >/dev/null 2>&1
if [ "$?" -eq 0 ];
then
  echo "CRITICAL - could not retrieve catalog from remote server |${PERFDATA}"
  exit 2
fi

LAST_RUN=$(grep last_run "${LAST_RUN_FILE}" 2>/dev/null | awk '{ print $NF }')

if [ -z "$LAST_RUN" ];
then
	echo "CRITICAL - error getting data from last_run_summary.yaml"
	exit 2
fi

NOW=$(date +%s)

DIFF_LAST_RUN=$(($NOW-$LAST_RUN))

if [ "$?" -ne 0 ];
then
  echo "CRITICAL - unable to fect last run data"
  exit 2
fi

if [ -z "${DIFF_LAST_RUN}" ];
then
  echo "CRITICAL - unable to fect last run data"
  exit 2
fi

PERFDATA="${PERFDATA}; difflastrun=${DIFF_LAST_RUN};"

if [ "${DIFF_LAST_RUN}" -ge "${MAXDIFF}" ];
then
	echo "CRITICAL - last run: ${DIFF_LAST_RUN} seconds ago |${PERFDATA}"
	exit 2
fi

if [ "${DIFF_LAST_RUN}" -ge "${WARNDIFF}" ];
then
  echo "WARNING - last run: ${DIFF_LAST_RUN} seconds ago |${PERFDATA}"
  exit 1
fi

# # grep -i catalo /var/lib/puppet/state/last_run_report.yaml
#       message: "Caching catalog for demovm.systemadmin.es"
#       message: "Finished catalog run in 10.89 seconds"
grep "Finished catalog run" "${LAST_RUN_REPORT}" >/dev/null 2>&1
if [ "$?" -eq 0 ];
then
  echo "OK - last run: ${DIFF_LAST_RUN} seconds ago |${PERFDATA}"
  exit 0
else
  grep "Applied catalog in " "${LAST_RUN_REPORT}" >/dev/null 2>&1
  if [ "$?" -eq 0 ];
  then
    echo "OK - last run: ${DIFF_LAST_RUN} seconds ago |${PERFDATA}"
    exit 0
  else
    echo "CRITICAL - puppet agent does not report to have finished to apply the catalog |$PERFDATA"
    exit 3
  fi
fi


files/puppet_last_run.sh
#!/bin/bash
# puppet managed file
# snmpd compatible check

PATH="/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/opt/puppetlabs/bin"

PUPPETBIN=$(which puppet 2>/dev/null)

if [ -z "${PUPPETBIN}" ];
then
  # puppet not found
  exit 2
fi

PUPPET_VER="$(${PUPPETBIN} --version 2>/dev/null)"

if [[ $PUPPET_VER = 3* ]];
then
  LAST_RUN_FILE='/var/lib/puppet/state/last_run_summary.yaml'
  LAST_RUN_REPORT='/var/lib/puppet/state/last_run_report.yaml'
elif [[ $PUPPET_VER = 5* ]];
then
  LAST_RUN_FILE='/opt/puppetlabs/puppet/cache/state/last_run_summary.yaml'
  LAST_RUN_REPORT='/opt/puppetlabs/puppet/cache/state/last_run_report.yaml'
else
  exit 3
fi

if [ ! -e "${LAST_RUN_FILE}" ];
then
	# last_run_summary.yaml does not exists
	exit 4
fi

if [ ! -e "${LAST_RUN_REPORT}" ];
then
	# last_run_report.yaml does not exists
	exit 5
fi

LAST_RUN=$(grep last_run ${LAST_RUN_FILE} 2>/dev/null | awk '{ print $NF }')

if [ -z "$LAST_RUN" ];
then
	# error getting data from last_run_summary.yaml
	exit 6
fi

#
# outputs
#
# # grep -i catalo /var/lib/puppet/state/last_run_report.yaml
#       message: "Could not retrieve catalog from remote server: Error 400 on SERVER: Puppet::Parser::AST::Resource failed with error ArgumentError: Could not find declared class sssd::monit at /etc/puppet/manifests/site.pp:919 on node centos7.systemadmin.es"
#       message: "Using cached catalog"
#       message: "Finished catalog run in 7.96 seconds"
#

grep "Using cached catalog" ${LAST_RUN_REPORT} >/dev/null 2>&1
if [ "$?" -eq 0 ];
then
  # server using cached catalog
  exit 7
fi

# # grep -i catalo /var/lib/puppet/state/last_run_report.yaml
#       message: "Could not retrieve catalog from remote server: Error 400 on SERVER: Puppet::Parser::AST::Resource failed with error ArgumentError: Invalid resource type pam::securetty at /etc/puppet/manifests/site.pp:356 on node testvm.systemadmin.es"
#       message: "Not using cache on failed catalog"
#       message: "Could not retrieve catalog; skipping run"
#
grep "Could not retrieve catalog from remote server" ${LAST_RUN_REPORT} >/dev/null 2>&1
if [ "$?" -eq 0 ];
then
  # could not retrieve catalog from remote server
  exit 8
fi

#Notice: Skipping run of Puppet configuration client; administratively disabled (Reason: 'reason not specified');
#Use 'puppet agent --enable' to re-enable.
tail -n 2 /var/log/puppet/puppet.log | grep disable >/dev/null 2>&1
if [ "$?" -eq 0 ];
then
  # minion is disabled
  exit 9
fi

# Puppet (err): Failed to apply catalog: no parameter
# Puppet catalog not being applied
tail -n 2 /var/log/puppet/puppet.log | grep "Failed to apply catalog" >/dev/null 2>&1
if [ "$?" -eq 0 ];
then
  # Catalog is not being applied
  exit 10
fi

NOW=$(date +%s)

DIFF_LAST_RUN=$(($NOW-$LAST_RUN))
RESOURCES="$(grep resources: ${LAST_RUN_FILE} -A7 | grep -v resources: | paste '-sd;' | sed -e 's/: /=/g' -e 's/^[^a-zA-Z]*//' -e 's/[ \t]+/ /g')"

echo "${RESOURCES}; difflastrun=${DIFF_LAST_RUN};"
exit 0
